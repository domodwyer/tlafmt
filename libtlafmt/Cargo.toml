[package]
name = "libtlafmt"
version = "0.2.0"
edition = "2021"
authors = ["Dom <dom@itsallbroken.com>"]
description = "A formatter library for TLA+ specs, core of tlafmt"
documentation = "https://docs.rs/tlafmt"
repository = "https://github.com/domodwyer/tlafmt"
license = "Apache-2.0"
categories = ["command-line-utilities", "development-tools"]
keywords = ["tla", "plus", "tla+", "formatter"]

[dependencies]
thiserror = "2.0.12"
tree-sitter = "0.25.3"
tree-sitter-tlaplus = "1.5.0"

[dev-dependencies]
insta = { version = "1.42.2", features = ["glob"] }
pretty_assertions = "1.4.1"

[lints]
workspace = true
