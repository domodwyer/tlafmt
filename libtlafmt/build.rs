#![allow(missing_docs)]

fn main() {
    // This statement prevents cargo from complaining about an unknown (to it)
    // cfg for fuzzing.
    println!("cargo::rustc-check-cfg=cfg(fuzzing)");
}
