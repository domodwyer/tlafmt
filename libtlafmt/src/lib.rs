mod ast_format;
mod helpers;
mod renderer;
mod test_utils;
mod token;

use std::io::Write;

use ast_format::format_node;
use helpers::EmptyLines;
use renderer::Renderer;
use thiserror::Error;
use tree_sitter::{Node, Parser, Tree};

const LINE_WIDTH: usize = 80;

/// Errors during AST parsing, lowering or rendering.
#[derive(Debug, Error)]
pub enum Error {
    /// An I/O error writing to the output sink.
    #[error("data store disconnected")]
    IO(#[from] std::io::Error),

    /// The input cannot be parsed into an AST.
    #[error("unknown parser error")]
    Parse,

    /// The module header is malformed.
    #[error("invalid module header")]
    ModuleHeader,

    /// A `[Next]_vars` sequence is malformed.
    #[error("invalid step-or-stutter sequence")]
    StepOrStutter,
}

/// A parsed TLA file.
///
/// Formatting a TLA file occurs in three phases:
///
///   1. The input file is parsed into an abstract syntax tree.
///   2. The AST is then lowered into a formatter-specific representation.
///   3. The format representation is rendered into output text.
///
/// Step (1) is performed when calling [`ParsedFile::new()`] to initialise a new
/// instance, and steps (2) and (3) are performed when [`ParsedFile::format()`]
/// is called.
#[derive(Debug)]
pub struct ParsedFile<'a> {
    t: Tree,
    input: &'a str,
}

impl<'a> ParsedFile<'a> {
    /// Parse the `input` TLA spec into an AST.
    pub fn new(input: &'a str) -> Result<Self, Error> {
        let mut parser = Parser::new();
        parser
            .set_language(&tree_sitter_tlaplus::LANGUAGE.into())
            .expect("error loading TLA+ grammar");

        Ok(Self {
            t: parser.parse(input, None).ok_or(Error::Parse)?,
            input,
        })
    }

    /// Format and render the parsed spec into `out`.
    ///
    /// # Errors
    ///
    /// If formatting fails `out` may contain partial content.
    pub fn format<W>(&self, out: W) -> Result<(), Error>
    where
        W: Write,
    {
        let mut out = Renderer::new(out);
        let mut empty_lines = EmptyLines::default();

        // Lower the AST into a series of formatter tokens wrote to `out`.
        format_node(self.t.root_node(), self.input, &mut empty_lines, &mut out)?;

        out.flush()?;

        Ok(())
    }
}

/// Return the content in `input` for `node.`
fn get_str<'a>(node: &Node<'_>, input: &'a str) -> &'a str {
    &input[node.byte_range()]
}

#[cfg(test)]
mod tests {
    use insta::glob;

    use super::*;

    /// Execute the formatter against each file in the TLA spec corpus, and
    /// compare the output against a reference copy.
    ///
    /// If this test fails, use "cargo insta review" to inspect any output
    /// changes.
    #[test]
    fn test_corpus() {
        use std::fs;

        glob!("../../", "tests/corpus/*.tla", |path| {
            let input = fs::read_to_string(path).expect("read test corpus file");
            assert_rewrite!(&input);
        });
    }
}
