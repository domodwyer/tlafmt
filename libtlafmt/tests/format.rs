//! Test formatting the embedded TLA+ corpus.

include!("../src/test_utils.rs"); // Pull in private assert_rewrite!

use thiserror as _;
use tree_sitter as _;
use tree_sitter_tlaplus as _;

use insta::glob;
use libtlafmt::ParsedFile;

/// Execute the formatter against each file in the TLA spec corpus, and compare
/// the output against a reference copy.
///
/// Additionally re-run the formatter against the output of the first run to
/// ensure stable output between repeated runs.
///
/// If this test fails, use "cargo insta review" to inspect any output changes.
#[test]
fn test_corpus() {
    use std::fs;

    glob!("../", "tests/corpus/*.tla", |path| {
        let input = fs::read_to_string(path).expect("read test corpus file");
        assert_rewrite!(&input);
    });
}
