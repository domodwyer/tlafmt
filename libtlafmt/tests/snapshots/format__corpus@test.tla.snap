---
source: libtlafmt/tests/format.rs
expression: output
input_file: libtlafmt/tests/corpus/test.tla
---
--------------------------------- MODULE main ----------------------------------

EXTENDS Integers, Sequences, TLC, FiniteSets

(*
--wf
--algorithm binary_search{
variables nums, n, target, low, high, mid;
{
    low := 0;
    high := n - 1;
    while (low <= high){

        mid := (low + high) / 2;

        if (target = nums[mid]){
            print mid;
        };
        if (target < nums[mid]) {
             high := mid - 1;
        } else{
            low := mid + 1;
        };


}
}}
}
end algorithm;

*)
\* BEGIN TRANSLATION (chksum(pcal) = "22207036" /\ chksum(tla) = "3c451e25")
CONSTANT defaultInitValue
VARIABLES nums, n, target, low, high, mid, pc

vars == << nums, n, target, low, high, mid, pc >>

Init == (* Global variables *)
    /\ nums = defaultInitValue
    /\ n = defaultInitValue
    /\ target = defaultInitValue
    /\ low = defaultInitValue
    /\ high = defaultInitValue
    /\ mid = defaultInitValue
    /\ pc = "Lbl_1"

Lbl_1 ==
    /\ pc = "Lbl_1"
    /\ low' = 0
    /\ high' = n - 1
    /\ pc' = "Lbl_2"
    /\ UNCHANGED << nums, n, target, mid >>

Lbl_2 ==
    /\ pc = "Lbl_2"
    /\ IF low <= high
        THEN
            /\ mid' = (low + high) / 2
            /\ IF target = nums[mid']
                THEN
                    /\ PrintT(mid')
                ELSE
                    /\ TRUE
            /\ IF target < nums[mid']
                THEN
                    /\ high' = mid' - 1
                    /\ low' = low
                ELSE
                    /\ low' = mid' + 1
                    /\ high' = high
            /\ pc' = "Lbl_2"
        ELSE
            /\ pc' = "Done"
            /\ UNCHANGED << low, high, mid >>
    /\ UNCHANGED << nums, n, target >>

(* Allow infinite stuttering to prevent deadlock on termination. *)
Terminating == pc = "Done" /\ UNCHANGED vars

Next == Lbl_1 \/ Lbl_2
\/ Terminating

Spec == Init /\ [][Next]_vars

Termination == <>(pc = "Done")

\* END TRANSLATION

================================================================================
